src/types/bytearray.rs

```diff
@@ -23,36 +23,55 @@ impl PyByteArray {
 
     /// Creates a new Python `bytearray` object with an `init` closure to write its contents.
     /// Before calling `init` the bytearray is zero-initialised.
+    /// * If `init` returns `Err(e)`, `new_with` will return `Err(e)`.
+    /// * If `init` returns `Ok(new_len)`, the allocated bytearray will be truncated to length
```

Comment:
Do we really want to truncate the bytes?
I cannot imagine any use case.

Reply:
We can easily remove the truncation again - it was mentioned by @davidhewitt for a potential API so I included it so we can see how it would look like. For the bytes, the idea is that as `init` implies that you are still generating the content, you might not know the exact size of the bytearray / bytestring up front, just an upper bound.

Reply:
Imagine you want to serialize some large dataset directly to Python. You may not know its final allocated length, so with this API you can pre-allocate the memory space in Python, write to it, and then return the amount of bytes written so that the array can be safely truncated.

I suppose that an alternative approach would be for argument to `init` to be a mutable reference to some custom struct instead of `&mut [u8]`:

```
pub struct UninitializedPyBytes<'a>(*mut ffi::PyObject, Python<'a>);

impl UninitializedPyBytes {
    /// Resize the byte buffer, zero-intializing any additional storage allocated
    /// May return Err on OOM
    pub fn resize(&mut self) -> PyResult<()> {  /* ... */  }

    /// Access the underlying bytes
    pub fn as_bytes(&mut self) -> &mut [u8] {  /* ... */  }
}
```

We could also implement `DerefMut<Target = [u8]>`, and maybe `Read` and `Write` for this. `Write` could even support automatically growing the allocation.

Reply:
Would we support the same interface for both `PyBytes` and `PyByteArray`? I don't think we need implement `io::Read` here, though. Regarding automatic resizing by `io::Write`, should we follow the simply size x 2 rule? Would we also automatically truncate the bytes after writing (which I found requires an additional struct that implements `Write`, references the bytes and implements `Drop` so we ensure that the truncation to the eventual length is performed after writing has completed)?

Reply:
Yep I guess `Read` is redundant. And yeah I was thinking the struct could have a private method which finishes the writing, truncates if needed, and returns a finished `PyBytes` object.

Reply:
Truncation has been removed in 4b3422e3b43d5fc43f430648f6479f1e2c4fcf18